contrib/gorilla/websocket: add tracing for websockets

Julio-Guerra opened the pull request:
- [x] Trace a websocket connection:  
   Trace the websocket connection as span `websocket.connection` from the call to function `websocket.Upgrade()` (which hijacks the HTTP request and returns the websocket connection `websocket.Conn`) to the call to method `(*websocket.Conn).Close()`.
- [x] Trace read and write methods of the websocket connection:
     - `ReadMessage()` as span `websocket.read_message`
     - `ReadJSON()` as span `websocket.read_json`
     - `WriteMessage()` as span `websocket.write_message`
     - `WriteJSON()` as span `websocket.write_json`
     - `WriteControl()` as span `websocket.write_control`
     - `WritePreparedMessage()` as span `websocket.write_prepared_message`
 - [ ] Tests

gbbr commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,47 @@
+
+			err = c.WriteMessage(mt, message)
+			if err != nil {
+				log.Println("write:", err)
```
```suggestion
				log.Printf("Error writing message: %v\n", err)
```

gbbr commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,47 @@
+		for {
+			mt, message, err := c.ReadMessage()
+			if err != nil {
+				log.Println("read:", err)
```
```suggestion
				log.Printf("Error reading message: %v", err)
```

gbbr commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,47 @@
+			mt, message, err := c.ReadMessage()
+			if err != nil {
+				log.Println("read:", err)
+				break
```
Aren't we supposed to continue and try the next message?

gbbr commented on contrib/gorilla/websocket/websocket.go:
```diff
@@ -0,0 +1,143 @@
+	"time"
+
+	"github.com/gorilla/websocket"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
```
We should create import groups (separated by an empty line).
1. Standard lib
2. Local imports
3. Other imports

Julio-Guerra replied:
Good to know, let me add this rule in your guide.

Julio-Guerra replied:
added https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#imports

gbbr commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,47 @@
+
+	var upgrader = websocketTrace.WrapUpgrader(websocket.Upgrader{})
+	mux.HandleFunc("/ws", func(w http.ResponseWriter, r *http.Request) {
+		c, err := upgrader.Upgrade(w, r, nil)
```
```suggestion
		conn, err := upgrader.Upgrade(w, r, nil)
```
Why not the common `conn`?

gbbr commented on contrib/gorilla/websocket/websocket.go:
```diff
@@ -0,0 +1,143 @@
+		}
+	}()
+
+	tracee, err := u.Upgrader.Upgrade(w, r.WithContext(ctx), responseHeader)
```
```suggestion
	conn, err := u.Upgrader.Upgrade(w, r.WithContext(ctx), responseHeader)
```
Why not `conn`?

Julio-Guerra replied:
the new WarpConn api now requires handling two conn variables in the same scope (unless writing some more line to have a separate scope - but good enough for this small example)

gbbr commented on contrib/gorilla/websocket/websocket.go:
```diff
@@ -0,0 +1,143 @@
+	}
+}
+
+// Close wraps the actual connection Close() method and finishes the
+// websocket.connection span started by Upgrade().
```
I think it's best to say what this does from a usage point of view. The detail here becomes confusing and seems to talk too much about the implementation. e.g. think more from a `godoc.org` perspective rather than `looking at code` perspective.
```suggestion
// Close closes the connection.
```
What do you think? I'd apply this logic everywhere.

Julio-Guerra replied:
It felt wrong to copy the original package documentation. But I noticed it indeed and was looking forward this comment ;-)

Happy to keep the current documentation approach 👍

gbbr commented on contrib/gorilla/websocket/websocket.go:
```diff
@@ -0,0 +1,143 @@
+// websocket.connection span started by Upgrade().
+func (c *Conn) Close() (err error) {
+	if span, ok := tracer.SpanFromContext(c.ctx); ok {
+		defer func() {
```
Very nice 👍🏻

gbbr commented on contrib/gorilla/websocket/websocket.go:
```diff
@@ -0,0 +1,143 @@
+// WriteControl traces the actual WriteControl() method using the global tracer
+// as span websocket.write_control.
+func (c *Conn) WriteControl(messageType int, data []byte, deadline time.Time) (err error) {
+	span, _ := tracer.StartSpanFromContext(c.ctx, "websocket.write_control",
+		messageTypeTag(messageType), messageLengthTag(len(data)))
```
I don't really like this. Can we drop the helper methods? I don't see any benefit in them, other than having a reader search for their definition to understand what they do here. Without them, even if more verbose, the code can be understood on the first pass because it's common, known API, and not newly introduced concepts. Basically what I'm saying is:
```suggestion
	span, _ := tracer.StartSpanFromContext(c.ctx, 
		"websocket.write_control",
		tracer.Tag("websocket.message_type", mtype),
		tracer.Tag("websocket.message_length", len(data)),
	)
```
Same below. And in the spirit of having short variable names can we use `mtype` here? I don't see a reason for a long descriptive name. In the spirit of "[naming things](https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#naming-things)".

Julio-Guerra replied:
Simple code factorization, in case the tag name had to be changed.

Julio-Guerra replied:
I changed it and inlined the tracer.Tag calls, factoring out the tag names into global const string values.

gbbr replied:
Perfect! THanks

gbbr commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,47 @@
+
+	"github.com/gorilla/websocket"
+	muxtrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/gorilla/mux"
+	websocketTrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/gorilla/websocket"
```
You resolved this but didn't change it.

gbbr replied:
☝🏻

Julio-Guerra replied:
my bad, I forgot that one

gbbr commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,48 @@
+func ExampleWrapConn() {
+	mux := muxtrace.NewRouter()
+
+	upgrader := websocket.Upgrader{}
```
Actually, the most common way to do this is probably:
```suggestion
	var upgrader websocket.Upgrader
```
Sorry for the back and forth

gbbr commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,48 @@
+			return
+		}
+		defer conn.Close()
+		tracedConn := websocketTrace.WrapConn(r.Context(), conn)
```
Can we find a way to call this `conn` ? Maybe the other one can be `oconn` (original connection)

gbbr commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+// This example illustrates the usage of WrapConn to trace the websockets
+// read and write operations.
```
```suggestion
// connections.
```

gbbr reviewed (commented):
You've requested a review, but there's a lot of comments still unaddressed

Julio-Guerra reviewed (commented):
Sorry about the glitches, I actually changed my local app example and not the example_test.go file 😅

felixge commented:
@Julio-Guerra if you're interested, I'd be happy to provide some review on this as well.

gbbr commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,48 @@
+	mux.HandleFunc("/ws", func(w http.ResponseWriter, r *http.Request) {
+		oconn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			log.Print("upgrade:", err)
```
```suggestion
			log.Printf("Error on upgrade: %v", err)
```
Just to be consistent with the other messages in the same example (capital letter at start, %v, etc)

Also, isn't the right approach here to call [`http.Error`](http://golang.org/pkg/net/http#Error) or something similar. We should promote good practices in our examples 🙂

gbbr commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,48 @@
+			}
+			log.Printf("Received message: %s\n", message)
+
+			err = conn.WriteMessage(mt, message)
+			if err != nil {
```
```suggestion
			if err := conn.WriteMessage(mt, message); err != nil {
```

gbbr commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,48 @@
+			}
+		}
+	})
+
```
Can we remove this line please? (see my style conventions doc)

Julio-Guerra replied:
are you against finishing text files with a newline?

Julio-Guerra replied:
FYI, I personally think code is more readable when split into logical blocks thanks to such newlines.
Similarly to the grouping you asked me to do with the imports.

gbbr commented on contrib/gorilla/websocket/websocket.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+const (
+	messageTypeTag   = "websocket.message_type"
+	messageLengthTag = "websocket.message_length"
```
We usually do it the other way around (`tag*`) shared prefix... Do you mind changing?

gbbr commented on contrib/gorilla/websocket/websocket.go:
```diff
@@ -0,0 +1,109 @@
+	messageLengthTag = "websocket.message_length"
+)
+
+// WrapConn wraps the websocket connection to trace its methods using the global
```
Wonderful 👍🏻

gbbr commented on contrib/gorilla/websocket/websocket.go:
```diff
@@ -0,0 +1,109 @@
+	span, _ := tracer.StartSpanFromContext(c.ctx,
+		"websocket.write_message",
+		tracer.Tag(messageTypeTag, messageType),
+		tracer.Tag(messageLengthTag, len(data)))
```
Nit: can you please put the closing parenthesis on the next line? (downstairs too)

Julio-Guerra replied:
In this case, I would also move the first argument `c.ctx` in its own line aligned with other arguments too, wdyt?

gbbr commented on contrib/gorilla/websocket/websocket_test.go:
```diff
@@ -0,0 +1,158 @@
+	}
+}
+
+func runClientServer(t *testing.T, client, server func(conn *websockettrace.Conn)) {
```
Sorry, but the fact that some code is a test helper, does not make it optional for testing. I think that in order to make sure our tests are accurate, it's mandatory to thoroughly test this function too, so I'd recommend having a `TestRunClientServer` bunch as well.

Julio-Guerra replied:
But the test would fail by design if this function doesn't work 🤔
For example the span assertions would fail in such case.

gbbr commented on contrib/gorilla/websocket/websocket_test.go:
```diff
@@ -0,0 +1,158 @@
+			},
+		},
+	} {
+		tc := tc
```
Let's remove this 🙂

Julio-Guerra replied:
Indeed we can avoid it because the test isn't run in parallel, so the following closure will be okay when synchronously executed.

But since this is a very common mistake that can be pretty hard to find, I tend to always define a scope there before having closures on them and avoid any issue.

Let me know what you really prefer =)

gbbr reviewed (commented):
This is really nice. Thanks a lot for taking the time. Just a few more things

felixge commented on contrib/gorilla/websocket/example_test.go:
```diff
@@ -0,0 +1,47 @@
+				log.Printf("Error writing message: %v\n", err)
+				continue
+			}
+		}
```
Is this a realistic use case?

If I was instrumenting a websocket application, I'd probably think of each `ReadMessage()` as a request that is answered by a follow-up `WriteMessage()`, like shown below:

```go
func wsHandler(w http.ResponseWriter, r *http.Request) {
	c, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Print("upgrade:", err)
		return
	}
	defer c.Close()

	for {
		txSpan := tracer.StartSpan("tx")
		func() {
			defer txSpan.Finish()

			readSpan := tracer.StartSpan("read", tracer.ChildOf(txSpan.Context()))
			mt, message, err := c.ReadMessage()
			if err != nil {
				readSpan.Finish(tracer.WithError(err))
				return
			}
			readSpan.Finish()

			processingSpan := tracer.StartSpan("processing", tracer.ChildOf(txSpan.Context()))
			time.Sleep(500 * time.Millisecond)
			processingSpan.Finish()

			writeSpan := tracer.StartSpan("write", tracer.ChildOf(txSpan.Context()))
			err = c.WriteMessage(mt, message)
			if err != nil {
				writeSpan.Finish(tracer.WithError(err))
				return
			}
			writeSpan.Finish()
		}()
	}
}
``` 

So IMO the API should somehow make this use case easier.

Wrapping the connection to track Read/Write durations by itself seems pretty uninformative, because it doesn't let us associate server-side processing time with the individual messages. Maybe a better API could look like this:

```go
func wsHandler2(w http.ResponseWriter, r *http.Request) {
	c, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Print("upgrade:", err)
		return
	}
	defer c.Close()

	for {
		tx := ddwebsocket.StartTx(c)
		func() {
			defer tx.Finish()

			mt, message, err := tx.ReadMessage() // automatically creates span
			if err != nil {
				return
			}

			processingSpan := tx.StartSpan("processing")
			time.Sleep(500 * time.Millisecond)
			processingSpan.Finish()

			err = tx.WriteMessage(mt, message) // // automatically creates span
			if err != nil {
				return
			}
		}()
	}
}
```

Anyway, I realize that this PR is perhaps more of an exercise than it is the solution to a particular problem, but for the sake of making this a realistic review, it'd be nice to clarify the use case for this so we can figure out if the API is addressing it well : ).

Julio-Guerra replied:
Good points indeed. The way the Go context is handled once the Conn created doesn't help (not exposed anymore, and maybe not even used). But I think you can actually do it by doing:

```golang
func wsHandler3(w http.ResponseWriter, r *http.Request) {
	oconn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Print("upgrade:", err)
		return
	}
	defer oconn.Close()

	// Create a tx span
	tx, ctx := tracer.StartFromContext(r.Context(), "tx")
	defer tx.Finish()
	// Trace the websocket connection under this tx span
	conn := websockettrace.WrapConn(ctx, oconn)
	for {
		mt, message, err := conn.ReadMessage() // automatically creates span
		if err != nil {
			return
		}

		processingSpan := tracer.StartSpanFromContext(ctx, "processing")
		time.Sleep(500 * time.Millisecond)
		processingSpan.Finish()

		err = conn.WriteMessage(mt, message) // // automatically creates span
		if err != nil {
			return
		}
	}
}
```

wdyt?

felixge replied:
> wdyt?

I was considering that route, but in your example you got a `tx` that has the same lifetime as the `oconn` connection itself. That might be fixable by moving both `WrapConn` as well as `StartFromContext` into the loop, but it seems weird to wrap the same connection multiple times inside of a loop. I mean it could probably work, but it seems really awkward from an API perspective. Do you agree?

Julio-Guerra replied:
I don't have enough background on integrations' "api style" to know if we sometimes provides such tracing helpers.
I'll let @gbbr tell us more about that.

felixge replied:
Sounds good. But just to confirm: You agree that the use case I outlined is reasonable and enabling it should be considered the key problem that this PR is trying to solve?

Julio-Guerra replied:
> Sounds good. But just to confirm: You agree that the use case I outlined is reasonable and enabling it should be considered the key problem that this PR is trying to solve?

Actually, the top-level websocket span (tx in our example) is going to be the same as the HTTP request span. So I am not so sure this one is useful. But the inner operation is for sure, and we definitely support this one.

felixge replied:
> Actually, the top-level websocket span (tx in our example) is going to be the same as the HTTP request span. 

Hm, but that could be a very long span with an unbound number of child spans. Didn't we discuss that this kind of thing doesn't work well with the tracer/agent as of today?

knusbaum replied:
Yes, it would be much better if we could separate individual requests/responses over the socket into individual traces. Having the whole connection's activity grouped into one trace is going to be mostly useless for any use case websockets is designed for.

Julio-Guerra replied:
Is this possible today?
I felt like the current tracing API misses the ability to "fork" a trace this such case, which would allow to have individual tracers for each sub-request.

felixge reviewed (commented):
PTAL at my comment. I've only tried to analyze the use cases for the API. Once it's clear what the use case is, I'll try provide more comments on the implementation details.

knusbaum reviewed (changes requested):
I think we need to consider @felixge's comment before merging this.

I think that having a websocket connection which handles many requests, possibly over many minutes, appear as one trace is not a good UX.

Julio-Guerra commented:
> @Julio-Guerra : Thanks for working on this. Is there anything that I can help with addressing comments in this PR? Once it's merged I would like to use this feature.

The solution isn't clear to me in the end. IIUC, @knusbaum, this implementation would keep the trace as long as the websocket connection stays alive. As a result, we were discussing [here](https://github.com/DataDog/dd-trace-go/pull/912#discussion_r623113034) what better API we could give to websocket users while avoiding that trace retention problem. But I don't have the solution in the end. Feel free to let us what examples of the comments are the closest to ours, it could really help. And maybe tell us what traces/spans would you like to get out of your websocket API :-)
